src/java.base/share/classes/jdk/crac/impl/OpenFilePolicies.java

```diff
@@ -0,0 +1,245 @@
+                if (entry.getKey().matches(p)) {
+                    return entry.getValue();
+                }
+            }
+        }
+        if ("fifo".equals(type) && fd > 2) {
+            return fifoPolicy;
+        }
+        return defaultPolicy;
+    }
+
+    /**
+     * Defines a treatment of file found open during checkpoint.
+     */
+    public enum BeforeCheckpoint {
+        /**
+         * The checkpoint fails with an appropriate error message. This is the
+         * default as it is safer to force applications handle the checkpoint,
+         * the options below are meant as workarounds when this is not feasible.
+         */
+        ERROR,
+        /**
+         * The file will be silently closed. The original path will
+         * be recorded and after restore this will be subject to treatment based
+         * on the {@link AfterRestore} policy.
+         */
+        CLOSE,
+        /**
+         * The behaviour is identical to {@link #CLOSE} but the application
+         * will print out a warning message to the standard error.
+         */
+        WARN_CLOSE,
+        // TODO: for no-downtime replication (scaling up) a strategy that would
+        //  keep the descriptor open when the checkpointed process is left
+        //  running might be useful.
+    }
+
+    /**
+     * Defines what to do with file descriptors closed by the
+     * {@link BeforeCheckpoint} policy after restore from a checkpoint.
+     */
+    public enum AfterRestore {
+        /**
+         * The file is reopened. If it cannot be opened (e.g. the file is not
+         * on the filesystem anymore or the process has insufficient
+         * permissions) an error is printed and the checkpoint throws
+         * an exception.
+         * This is the default behaviour.
+         */
+        REOPEN_OR_ERROR,
+        /**
+         * The file is reopened. If it cannot be opened (e.g. the file
+         * is missing or the process has insufficient permissions) a /dev/null
+         * device is opened instead (allowing the process to write any data and
+         * immediately returning EOF when reading).
+         */
+        REOPEN_OR_NULL,
+        /**
+         * The file is reopened, ignoring the last offset and using the end
+         * of file instead. This is particularly useful for append-only logs.
+         * If it cannot be reopened the behaviour is identical to {@link #REOPEN_OR_ERROR}.
+         */
+        REOPEN_AT_END,
+        /**
+         * After restore another file specified as part of the policy
+         * declaration is opened instead; the policy name should be followed
+         * by an equal sign '=' and the new path.
+         * If the other file cannot be open the checkpoint throws an exception.
+         */
+        OPEN_OTHER,
+        /**
+         * Similar to {@link #OPEN_OTHER} but ignored the previous offset
+         * and opens the file at the end.
+         */
+        OPEN_OTHER_AT_END,
```

Comment:
Semantically, these modes are the same, the second is for FD opened in APPEND mode. But having both modes allows changing the FD mode for the application. Such change will break app assumptions about the file descriptor.

Reply:
There's many ways that you could break an application invariant if the file is changed/removed when the application thinks it has exclusive write access to it, e.g. it might not expect to write files bigger than X because it streams at most X bytes to that, but if you change the position during reopening you might violate that expectation. These are different options to handle that case: if you think the simplicity of use outweighs the choice I can make OPEN_OTHER use the _AT_END semantics when the file was opened for appending.